service/labelstore/service.go

```diff
@@ -0,0 +1,226 @@
+package labelstore
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	agent_service "github.com/grafana/agent/service"
+	flow_service "github.com/grafana/agent/service"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const ServiceName = "labelstore"
+
+type service struct {
+	log                log.Logger
+	mut                sync.Mutex
+	globalRefID        uint64
+	mappings           map[string]*remoteWriteMapping
+	labelsHashToGlobal map[uint64]uint64
+	staleGlobals       map[uint64]*staleMarker
+}
+
+type staleMarker struct {
+	globalID        uint64
+	lastMarkedStale time.Time
+	labelHash       uint64
+}
+
+type Arguments struct{}
+
+var _ flow_service.Service = (*service)(nil)
+
+func New(l log.Logger) *service {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+	return &service{
+		log:                l,
+		globalRefID:        0,
+		mappings:           make(map[string]*remoteWriteMapping),
+		labelsHashToGlobal: make(map[uint64]uint64),
+		staleGlobals:       make(map[uint64]*staleMarker),
+	}
+}
+
+// Definition returns the Definition of the Service.
+// Definition must always return the same value across all
+// calls.
+func (s *service) Definition() agent_service.Definition {
+	return agent_service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil,
+	}
+}
+
+// Run starts a Service. Run must block until the provided
+// context is canceled. Returning an error should be treated
+// as a fatal error for the Service.
+func (s *service) Run(ctx context.Context, host agent_service.Host) error {
+	<-ctx.Done()
+	return nil
+}
+
+// Update updates a Service at runtime. Update is never
+// called if [Definition.ConfigType] is nil. newConfig will
+// be the same type as ConfigType; if ConfigType is a
+// pointer to a type, newConfig will be a pointer to the
+// same type.
+//
+// Update will be called once before Run, and may be called
+// while Run is active.
+func (s *service) Update(newConfig any) error {
+	return nil
+}
+
+// Data returns the Data associated with a Service. Data
+// must always return the same value across multiple calls,
+// as callers are expected to be able to cache the result.
+//
+// Data may be invoked before Run.
+func (s *service) Data() any {
+	return s
+}
+
+// GetOrAddLink is called by a remote_write endpoint component to add mapping and get back the global id.
+func (s *service) GetOrAddLink(componentID string, localRefID uint64, lbls labels.Labels) uint64 {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	// If the mapping doesn't exist then we need to create it
+	m, found := s.mappings[componentID]
+	if !found {
+		m = &remoteWriteMapping{
+			RemoteWriteID: componentID,
+			localToGlobal: make(map[uint64]uint64),
+			globalToLocal: make(map[uint64]uint64),
+		}
+		s.mappings[componentID] = m
+	}
+
+	labelHash := lbls.Hash()
+	globalID, found := s.labelsHashToGlobal[labelHash]
```

Comment:
Does this handle hash collisions?

Reply:
It does not but its what we have been running for a year or so.